### PR TITLE
Number input range min max limits not working as expected

### PR DESCRIPTION
**Pull request by gayanSandamal**
#### Min / Max limitation and Step

https://github.com/user-attachments/assets/4900230e-bc74-4d77-8150-a174f39b371a

#### Responsiveness

Note that I've hide icon and clear button when widget gets shrink and there is no room for the payload to display

https://github.com/user-attachments/assets/1f2dd84e-882b-48bf-a095-3adfedcdeaca

## Related Issue(s)

https://github.com/FlowFuse/node-red-dashboard/issues/1244

## Checklist

 - [x] I have read the [contribution guidelines](https://github.com/FlowFuse/flowfuse/blob/main/CONTRIBUTING.md)
 - [ ] Suitable unit/system level tests have been added and they pass 
 - [x] Documentation has been updated
    - [ ] Upgrade instructions
    - [ ] Configuration details
    - [ ] Concepts
 - [ ] Changes `flowforge.yml`?
    - [ ] Issue/PR raised on `FlowFuse/helm` to update ConfigMap Template
    - [ ] Issue/PR raised on `FlowFuse/CloudProject` to update values for Staging/Production

## Labels

 - [ ] Includes a DB migration? -> add the `area:migration` label

### CODE REVIEW

**Comment by Steve-Mcl** [comment]
@gayanSandamal can you also include built-in help update in `nodes/widgets/locales/en-US/ui_number_input.html`.
Please ensure all properties that support dynamic setting (either existing or added in this PR) are listed.
Ta.

**Review by Steve-Mcl — CHANGES_REQUESTED** [review]
This PR adds dynamic support for min, max, step but the nodes built-in help is not updated.

Please update `nodes/widgets/locales/en-US/ui_number_input.html` to reflect the dynamic support

**Review by Steve-Mcl — CHANGES_REQUESTED** [review]
Apart from the suggested text update to say "`msg.ui_update`" I noticed this...

https://github.com/FlowFuse/node-red-dashboard/pull/1273/files#diff-6c3a8b98faadfc018feafc6225a11d60436505ac049ee150485c575cdfc29e4eR192-R194

I dont think the `onChange` ↑ handler will work as expected when the `step` is > 1.

File: nodes/widgets/locales/en-US/ui_number_input.html [hunk 1]
```diff
@@ -6,23 +6,30 @@ <h3>Dynamic Properties (Inputs)</h3>
     <p>Any of the following can be appended to a <code>msg.</code> in order to override or set properties on this node at runtime.</p>
```
[the remark this answers is not in the document]
Comment (gayanSandamal) [reply nodes/widgets/locales/en-US/ui_number_input.html:6]: @Steve-Mcl you're correct I've updated the onChange logic [here](https://github.com/FlowFuse/node-red-dashboard/pull/1273/commits/fa699ac31734cbbb211c3c9051cd27f42a05bba0)

regarding the docs update I've noticed the same in [core-widgets.md](https://github.com/FlowFuse/node-red-dashboard/blob/a371146f7ac8a7c36c27714eb6a39ec214af69ff/docs/contributing/widgets/core-widgets.md?plain=1#L255)

Screenshot 2024-09-11 at 10 04 15

**Comment by joepavitt** [comment]
Screenshot 2024-09-11 at 11 50 29

Tried setting the step and min/max and get errors

**Comment by joepavitt** [comment]
Enforcing the `controlVariant` to be stacked too isn't friendly for touchscreens, so please make this a configurable option for the Number Input itself

**Comment by Steve-Mcl** [comment]
Additional issues...

1. Setting a negative step permits number to escape bounds (top number input in demo gif)
   1. Result is unpredictable. Either we support negative spin or add validation  
   3. Operating the spinner button permits the value to escape min/max
2. In some cases, operating keyboard up/down not possible (think it is related to step > 1, not sure)
3. In some cases, the spinner buoon is greyed out (see 4th spinner in gif) - max is 100 but spinner button is greyed


**IMPORTANT CLUE**
After injecting dynamic values for all props, everything appears to work correctly (except negative step)


Demo:
chrome_dxBwBqVafK


Demo flow:
```
[{"id":"cbdcfc6a2166b7ee","type":"ui-number-input","z":"f80624abfd69b02e","group":"d8d5ce268004b131","name":"","label":"auto, min:0, max:10, step 1","order":14,"width":0,"height":0,"topic":"topic","topicType":"msg","min":0,"max":10,"step":1,"tooltip":"","passthru":true,"sendOnBlur":true,"sendOnEnter":true,"className":"","clearable":false,"icon":"","iconPosition":"left","iconInnerPosition":"inside","x":2280,"y":1200,"wires":[[]]},{"id":"b04429265d510afe","type":"ui-number-input","z":"f80624abfd69b02e","group":"d8d5ce268004b131","name":"","label":"2x1, min 0, max 10, step 1","order":16,"width":"2","height":"1","topic":"topic","topicType":"msg","min":0,"max":10,"step":1,"tooltip":"","passthru":true,"sendOnBlur":true,"sendOnEnter":true,"className":"","clearable":false,"icon":"","iconPosition":"left","iconInnerPosition":"inside","x":2280,"y":1240,"wires":[[]]},{"id":"f977fa12fa24d0e9","type":"ui-number-input","z":"f80624abfd69b02e","group":"d8d5ce268004b131","name":"","label":"6x1, min 0, max 9, step 3, icon+tooltip+clear","order":13,"width":"6","height":"1","topic":"topic","topicType":"msg","min":0,"max":"9","step":"3","tooltip":"6x1, min 0, max 9, step 3, icon+tooltip+clear","passthru":true,"sendOnBlur":true,"sendOnEnter":true,"className":"","clearable":true,"icon":"cog","iconPosition":"left","iconInnerPosition":"inside","x":2330,"y":1280,"wires":[[]]},{"id":"4936b7f3448d30e8","type":"ui-number-input","z":"f80624abfd69b02e","group":"d8d5ce268004b131","name":"","label":"6x1, min 0, max 15, step -5, icon+tooltip+clear","order":12,"width":"6","height":"1","topic":"topic","topicType":"msg","min":0,"max":"15","step":"-5","tooltip":"6x1, min 0, max 15, step -5, icon+tooltip+clear","passthru":true,"sendOnBlur":true,"sendOnEnter":true,"className":"","clearable":true,"icon":"home","iconPosition":"left","iconInnerPosition":"outside","x":2340,"y":1320,"wires":[[]]},{"id":"bc628c01a00bc840","type":"inject","z":"f80624abfd69b02e","name":"dynamic","props":[{"p":"ui_update.step","v":"2","vt":"num"},{"p":"ui_update.min","v":"4","vt":"num"},{"p":"ui_update.max","v":"8","vt":"num"},{"p":"ui_update.icon","v":"check","vt":"str"},{"p":"ui_update.label","v":"fred","vt":"str"},{"p":"ui_update.clearable","v":"true","vt":"bool"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1960,"y":1220,"wires":[["c68b54d25a2dd6c8"]]},{"id":"c5c577944d07723b","type":"inject","z":"f80624abfd69b02e","name":"reset dynamic","props":[{"p":"ui_update.step","v":"null","vt":"json"},{"p":"ui_update.min","v":"null","vt":"json"},{"p":"ui_update.max","v":"null","vt":"json"},{"p":"ui_update.icon","v":"null","vt":"json"},{"p":"ui_update.label","v":"null","vt":"json"},{"p":"ui_update.clearable","v":"null","vt":"json"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1986,"y":1302,"wires":[["c68b54d25a2dd6c8"]]},{"id":"2394c0ef8e1b0431","type":"ui-number-input","z":"f80624abfd69b02e","group":"d8d5ce268004b131","name":"","label":"auto, min 0, max 100, step 1","order":15,"width":0,"height":0,"topic":"topic","topicType":"msg","min":0,"max":"100","step":"1","tooltip":"Hello tooltip","passthru":true,"sendOnBlur":true,"sendOnEnter":true,"className":"","clearable":false,"icon":"","iconPosition":"left","iconInnerPosition":"inside","x":2280,"y":1380,"wires":[[]]},{"id":"c68b54d25a2dd6c8","type":"junction","z":"f80624abfd69b02e","x":2080,"y":1220,"wires":[["cbdcfc6a2166b7ee","b04429265d510afe","f977fa12fa24d0e9","4936b7f3448d30e8","2394c0ef8e1b0431"]]},{"id":"d8d5ce268004b131","type":"ui-group","name":"Dynamic Demo","page":"1525cc4810388af6","width":"12","height":"1","order":1,"showTitle":true,"className":"","visible":"true","disabled":"false"},{"id":"1525cc4810388af6","type":"ui-page","name":"Dynamic Demo","ui":"22ea43815413e748","path":"/mssql","icon":"home","layout":"grid","theme":"c2ff5ba1f92a0f0e","order":1,"className":"","visible":"true","disabled":"false"},{"id":"22ea43815413e748","type":"ui-base","name":"base","path":"/dashboard","includeClientData":true,"acceptsClientConfig":["ui-notification","ui-control"],"showPathInSidebar":false,"showPageTitle":true,"navigationStyle":"default","titleBarStyle":"default"},{"id":"c2ff5ba1f92a0f0e","type":"ui-theme","name":"Default","colors":{"surface":"#ffffff","primary":"#813b5a","bgPage":"#beacac","groupBg":"#ffffff","groupOutline":"#cc2828"},"sizes":{"pagePadding":"12px","groupGap":"12px","groupBorderRadius":"4px","widgetGap":"12px","density":"default"}}]
```

**Comment by gayanSandamal** [comment]
> Enforcing the `controlVariant` to be stacked too isn't friendly for touchscreens, so please make this a configurable option for the Number Input itself

I discussed with @Steve-Mcl last week and provided this stacked control view as the native input type["number"] has this stacked layout by default and the input area got covered by the controllers on 1x1 size.

But as you mentioned it's not user friendly for touchscreens. I will add this as a configuration and also mention that the "_stacked layout is recommended for smaller grid sizes_"

**Comment by gayanSandamal** [comment]
> Additional issues...
> 
> 1. Setting a negative step permits number to escape bounds (top number input in demo gif)
>    
>    1. Result is unpredictable. Either we support negative spin or add validation
>    2. Operating the spinner button permits the value to escape min/max
> 2. In some cases, operating keyboard up/down not possible (think it is related to step > 1, not sure)
> 3. In some cases, the spinner buoon is greyed out (see 4th spinner in gif) - max is 100 but spinner button is greyed
> 
> **IMPORTANT CLUE** After injecting dynamic values for all props, everything appears to work correctly (except negative step)

Thanks @Steve-Mcl for the detailed feedback! I’ll investigate with the NR dashboard flow given and fix the issues you pointed and the remaining edge cases as well.

**Comment by gayanSandamal** [comment]
> Enforcing the `controlVariant` to be stacked too isn't friendly for touchscreens, so please make this a configurable option for the Number Input itself

Added a new property called "Spinner" by giving the default value as "Inline" which is touch screen friendly
Screenshot 2024-09-12 at 13 27 25

But it will lead to this cluttered layout in 1x1 grid size


Therefore the user can change the spinner type to "Stacked" if necessary

**Comment by gayanSandamal** [comment]
> Additional issues...
> 
> 1. Setting a negative step permits number to escape bounds (top number input in demo gif)
>    
>    1. Result is unpredictable. Either we support negative spin or add validation
>    2. Operating the spinner button permits the value to escape min/max
> 2. In some cases, operating keyboard up/down not possible (think it is related to step > 1, not sure)
> 3. In some cases, the spinner buoon is greyed out (see 4th spinner in gif) - max is 100 but spinner button is greyed
> 
> **IMPORTANT CLUE** After injecting dynamic values for all props, everything appears to work correctly (except negative step)
> 
> Demo: chrome_dxBwBqVafK [ chrome_dxBwBqVafK ](https://private-user-images.githubusercontent.com/44235289/366428457-2faeb6a6-bdb0-48df-b42f-a930edcd02ef.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYxMjkxMTcsIm5iZiI6MTcyNjEyODgxNywicGF0aCI6Ii80NDIzNTI4OS8zNjY0Mjg0NTctMmZhZWI2YTYtYmRiMC00OGRmLWI0MmYtYTkzMGVkY2QwMmVmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEyVDA4MTMzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRjOTJlNTIxN2VlNTY5OTNkMGQzYjJiMTY3N2Q2NTRjNjFmZDljMmUxNDExMzA0MzI0YjJlZjM0Njk5MDNjYTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NDXnFfG6t0jTvQRqH_VH2R-EYCzDOjxxY2xnGsV-Jns) [ ](https://private-user-images.githubusercontent.com/44235289/366428457-2faeb6a6-bdb0-48df-b42f-a930edcd02ef.gif?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3MjYxMjkxMTcsIm5iZiI6MTcyNjEyODgxNywicGF0aCI6Ii80NDIzNTI4OS8zNjY0Mjg0NTctMmZhZWI2YTYtYmRiMC00OGRmLWI0MmYtYTkzMGVkY2QwMmVmLmdpZj9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MTIlMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTEyVDA4MTMzN1omWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWRjOTJlNTIxN2VlNTY5OTNkMGQzYjJiMTY3N2Q2NTRjNjFmZDljMmUxNDExMzA0MzI0YjJlZjM0Njk5MDNjYTYmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0JmFjdG9yX2lkPTAma2V5X2lkPTAmcmVwb19pZD0wIn0.NDXnFfG6t0jTvQRqH_VH2R-EYCzDOjxxY2xnGsV-Jns)
> 
> Demo flow:
> 
> ```
> [{"id":"cbdcfc6a2166b7ee","type":"ui-number-input","z":"f80624abfd69b02e","group":"d8d5ce268004b131","name":"","label":"auto, min:0, max:10, step 1","order":14,"width":0,"height":0,"topic":"topic","topicType":"msg","min":0,"max":10,"step":1,"tooltip":"","passthru":true,"sendOnBlur":true,"sendOnEnter":true,"className":"","clearable":false,"icon":"","iconPosition":"left","iconInnerPosition":"inside","x":2280,"y":1200,"wires":[[]]},{"id":"b04429265d510afe","type":"ui-number-input","z":"f80624abfd69b02e","group":"d8d5ce268004b131","name":"","label":"2x1, min 0, max 10, step 1","order":16,"width":"2","height":"1","topic":"topic","topicType":"msg","min":0,"max":10,"step":1,"tooltip":"","passthru":true,"sendOnBlur":true,"sendOnEnter":true,"className":"","clearable":false,"icon":"","iconPosition":"left","iconInnerPosition":"inside","x":2280,"y":1240,"wires":[[]]},{"id":"f977fa12fa24d0e9","type":"ui-number-input","z":"f80624abfd69b02e","group":"d8d5ce268004b131","name":"","label":"6x1, min 0, max 9, step 3, icon+tooltip+clear","order":13,"width":"6","height":"1","topic":"topic","topicType":"msg","min":0,"max":"9","step":"3","tooltip":"6x1, min 0, max 9, step 3, icon+tooltip+clear","passthru":true,"sendOnBlur":true,"sendOnEnter":true,"className":"","clearable":true,"icon":"cog","iconPosition":"left","iconInnerPosition":"inside","x":2330,"y":1280,"wires":[[]]},{"id":"4936b7f3448d30e8","type":"ui-number-input","z":"f80624abfd69b02e","group":"d8d5ce268004b131","name":"","label":"6x1, min 0, max 15, step -5, icon+tooltip+clear","order":12,"width":"6","height":"1","topic":"topic","topicType":"msg","min":0,"max":"15","step":"-5","tooltip":"6x1, min 0, max 15, step -5, icon+tooltip+clear","passthru":true,"sendOnBlur":true,"sendOnEnter":true,"className":"","clearable":true,"icon":"home","iconPosition":"left","iconInnerPosition":"outside","x":2340,"y":1320,"wires":[[]]},{"id":"bc628c01a00bc840","type":"inject","z":"f80624abfd69b02e","name":"dynamic","props":[{"p":"ui_update.step","v":"2","vt":"num"},{"p":"ui_update.min","v":"4","vt":"num"},{"p":"ui_update.max","v":"8","vt":"num"},{"p":"ui_update.icon","v":"check","vt":"str"},{"p":"ui_update.label","v":"fred","vt":"str"},{"p":"ui_update.clearable","v":"true","vt":"bool"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1960,"y":1220,"wires":[["c68b54d25a2dd6c8"]]},{"id":"c5c577944d07723b","type":"inject","z":"f80624abfd69b02e","name":"reset dynamic","props":[{"p":"ui_update.step","v":"null","vt":"json"},{"p":"ui_update.min","v":"null","vt":"json"},{"p":"ui_update.max","v":"null","vt":"json"},{"p":"ui_update.icon","v":"null","vt":"json"},{"p":"ui_update.label","v":"null","vt":"json"},{"p":"ui_update.clearable","v":"null","vt":"json"}],"repeat":"","crontab":"","once":false,"onceDelay":0.1,"topic":"","x":1986,"y":1302,"wires":[["c68b54d25a2dd6c8"]]},{"id":"2394c0ef8e1b0431","type":"ui-number-input","z":"f80624abfd69b02e","group":"d8d5ce268004b131","name":"","label":"auto, min 0, max 100, step 1","order":15,"width":0,"height":0,"topic":"topic","topicType":"msg","min":0,"max":"100","step":"1","tooltip":"Hello tooltip","passthru":true,"sendOnBlur":true,"sendOnEnter":true,"className":"","clearable":false,"icon":"","iconPosition":"left","iconInnerPosition":"inside","x":2280,"y":1380,"wires":[[]]},{"id":"c68b54d25a2dd6c8","type":"junction","z":"f80624abfd69b02e","x":2080,"y":1220,"wires":[["cbdcfc6a2166b7ee","b04429265d510afe","f977fa12fa24d0e9","4936b7f3448d30e8","2394c0ef8e1b0431"]]},{"id":"d8d5ce268004b131","type":"ui-group","name":"Dynamic Demo","page":"1525cc4810388af6","width":"12","height":"1","order":1,"showTitle":true,"className":"","visible":"true","disabled":"false"},{"id":"1525cc4810388af6","type":"ui-page","name":"Dynamic Demo","ui":"22ea43815413e748","path":"/mssql","icon":"home","layout":"grid","theme":"c2ff5ba1f92a0f0e","order":1,"className":"","visible":"true","disabled":"false"},{"id":"22ea43815413e748","type":"ui-base","name":"base","path":"/dashboard","includeClientData":true,"acceptsClientConfig":["ui-notification","ui-control"],"showPathInSidebar":false,"showPageTitle":true,"navigationStyle":"default","titleBarStyle":"default"},{"id":"c2ff5ba1f92a0f0e","type":"ui-theme","name":"Default","colors":{"surface":"#ffffff","primary":"#813b5a","bgPage":"#beacac","groupBg":"#ffffff","groupOutline":"#cc2828"},"sizes":{"pagePadding":"12px","groupGap":"12px","groupBorderRadius":"4px","widgetGap":"12px","density":"default"}}]
> ```

I applied the necessary fixes and this works better now. Also note that I've validated negative integers for steps

https://github.com/user-attachments/assets/823b6113-3637-4426-a4cc-07c6950bc287

**Comment by gayanSandamal** [comment]
> Screenshot 2024-09-11 at 11 50 29
> Tried setting the step and min/max and get errors, only the last error here is relevant

This is also handled with https://github.com/FlowFuse/node-red-dashboard/pull/1273/commits/cec107c7221e5d502d35647a33fe833c7bb8e119

**Comment by gayanSandamal** [comment]
> @gayanSandamal does this mean that you consider this is now ready for re-review?

@joepavitt Yes, this is ready for review

**Review by joepavitt — APPROVED** [review]

**Comment by joepavitt** [comment]
Have gone through all of the comments and feedback here, and as far as I can tell, it's all working as planned.

**Pull request merged**